### Dataset

In this homework, we'll build a model for predicting if we have an image of a dog or a cat. For this, we will use the "Dogs & Cats" dataset that can be downloaded from Kaggle.

You need to download the train.zip file.

If you have troubles downloading from Kaggle, use this link instead:

wget https://github.com/alexeygrigorev/large-datasets/releases/download/dogs-cats/train.zip

In the lectures we saw how to use a pre-trained neural network. In the homework, we'll train a much smaller model from scratch.



In [25]:
! rm -rf train.zip 

In [26]:
! rm -rf train/

In [27]:
!rm -rf /content/train_data

In [28]:
!rm -rf /content/val_data

In [29]:
ls 

sample_data/


In [30]:
!wget https://github.com/alexeygrigorev/large-datasets/releases/download/dogs-cats/train.zip

--2021-11-20 00:26:53--  https://github.com/alexeygrigorev/large-datasets/releases/download/dogs-cats/train.zip
Resolving github.com (github.com)... 13.114.40.48
Connecting to github.com (github.com)|13.114.40.48|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github-releases.githubusercontent.com/426348925/f39169c9-5f22-4a57-bb37-495c0d2974ab?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20211120%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20211120T002653Z&X-Amz-Expires=300&X-Amz-Signature=ad6bcad4f7db44b0ee2d75eb8f650946760c9e8044a86eb41295c35fd618eafd&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=426348925&response-content-disposition=attachment%3B%20filename%3Dtrain.zip&response-content-type=application%2Foctet-stream [following]
--2021-11-20 00:26:53--  https://github-releases.githubusercontent.com/426348925/f39169c9-5f22-4a57-bb37-495c0d2974ab?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVE

### Data Preparation

The dataset contains 12,500 images of cats and 12,500 images of dogs.

Now we need to split this data into train and validation

Create a train and validation folders

In each folder, create cats and dogs folders
Move the first 10,000 images to the train folder (from 0 to 9999) for boths cats and dogs - and put them in respective folders

Move the remaining 2,500 images to the validation folder (from 10000 to 12499)
You can do this manually or with Python (check os and shutil packages)

In [31]:
import shutil
shutil.unpack_archive("train.zip","/content/") #unzip train.zip

In [32]:
# check directory content
!ls -l  train/ | wc -l

25001


In [33]:
! ls train/ | grep "dog" | wc -l

12500


In [34]:
! ls train/ | grep "cat" | wc -l

12500


In [35]:
#create two folders train_data and val_data
import shutil, os
# Parent Directory path
parent_dir = "/content/"
folders = ['train_data','val_data']
for folder in folders:
  path = parent_dir+folder
  os.mkdir(path)



In [36]:
folders = ['/content/train_data/','/content/val_data/']
sub_folders = ['cats','dogs']
for folder in folders:
  for subfolder in sub_folders:
    path = folder+subfolder
    os.mkdir(path)

In [37]:
!ls train_data/

cats  dogs


In [38]:
!ls val_data/

cats  dogs


In [39]:
source_dir = "/content/train"
   
file_names = os.listdir(source_dir)

cat_count=0
dog_count=0

for file_name in file_names:
    
      if "cat" in file_name:
          cat_count+=1
          if (cat_count <= 10000):
             target_dir = '/content/train_data/cats'
             shutil.move(os.path.join(source_dir, file_name), target_dir)
          else:
             target_dir = '/content/val_data/cats'
             shutil.move(os.path.join(source_dir, file_name), target_dir)

      elif "dog" in file_name:
          dog_count+=1
          if (dog_count <= 10000):
            target_dir = '/content/train_data/dogs'
            shutil.move(os.path.join(source_dir, file_name), target_dir)
          else:
            target_dir = '/content/val_data/dogs'
            shutil.move(os.path.join(source_dir, file_name), target_dir)



In [40]:
!ls train_data/cats | wc -l

10000


In [41]:
!ls train_data/dogs | wc -l

10000


In [42]:
!ls val_data/cats | wc -l

2500


In [43]:
!ls val_data/dogs | wc -l

2500


Now we have 10000 images of cats and dogs in train_data and 2500 each of dogs and cats in val_data as a part of data preparation process

## Model

For this homework we will use Convolutional Neural Network (CNN. Like in the lectures, we'll use Keras.

You need to develop the model with following structure:

The shape for input should be (150, 150, 3)
Next, create a covolutional layer (Conv2D):
Use 32 filters
Kernel size should be (3, 3) (that's the size of the filter)
Use 'relu' as activation
Reduce the size of the feature map with max pooling (MaxPooling2D)
Set the pooling size to (2, 2)
Turn the multi-dimensional result into vectors using a Flatten layer
Next, add a Dense layer with 64 neurons and 'relu' activation
Finally, create the Dense layer with 1 neuron - this will be the output
The output layer should have an activation - use the appropriate activation for the binary classification case
As optimizer use SGD with the following parameters:

SGD(lr=0.002, momentum=0.8)


In [44]:
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.preprocessing.image import load_img
from tensorflow.keras.preprocessing.image import ImageDataGenerator

We’ll first add a convolutional 2D layer with 32 filters, a kernel of 3x3, the input size as our image dimensions, 150x150x3, and the activation as ReLU.

add a max pooling layer that halves the image dimension, so after this layer, the output will be 75x75x3

In [45]:
model = tf.keras.models.Sequential([
# Note the input shape is the desired size of the image 150x150 with 3 bytes color
# This is the first convolution
        tf.keras.layers.Conv2D(32, (3,3), activation='relu', input_shape=(150,150,3)),
        tf.keras.layers.MaxPooling2D(2, 2),

# Flatten the results to feed into a DNN
      tf.keras.layers.Flatten(),
# 64 neuron hidden layer
      tf.keras.layers.Dense(64, activation='relu'),
# Only 1 output neuron. It will contain a value from 0-1 where 0 for 1 class ('cat') and 1 for the other ('dog')
      tf.keras.layers.Dense(1, activation='sigmoid')
])

In [46]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 148, 148, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 74, 74, 32)       0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 175232)            0         
                                                                 
 dense (Dense)               (None, 64)                11214912  
                                                                 
 dense_1 (Dense)             (None, 1)                 65        
                                                                 
Total params: 11,215,873
Trainable params: 11,215,873
Non-trainable params: 0
____________________________________________

Question 1
Since we have a binary classification problem, what is the best loss function for us?

Note: since we specify an activation for the output layer, we don't need to set from_logits=True

Ans **binary_crossentropy**

Question 2


What's the total number of parameters of the model? You can use the summary method for that.

Ans **11,215,873**



In [47]:
optimizer = tf.keras.optimizers.SGD(
                                    learning_rate=0.002, momentum=0.8)

model.compile(loss='binary_crossentropy',
              optimizer=optimizer,
              metrics=['accuracy'])

In [48]:
train_gen = ImageDataGenerator(rescale=1./255)

train_ds = train_gen.flow_from_directory(
                       '/content/train_data',
                        target_size=(150, 150),
                        batch_size=20,
                        class_mode = 'binary'
)

Found 20000 images belonging to 2 classes.


In [49]:
train_ds.class_indices

{'cats': 0, 'dogs': 1}

In [50]:
val_gen = ImageDataGenerator(rescale=1./255)

val_ds = val_gen.flow_from_directory(
              '/content/val_data',
              target_size=(150, 150),
              batch_size=20,
              class_mode = 'binary'
)

Found 5000 images belonging to 2 classes.


In [51]:
train_ds.class_mode

'binary'

In [52]:
history = model.fit(
    train_ds,
    steps_per_epoch=100,
    epochs=10,
    validation_data=val_ds,
    validation_steps=50
)

Epoch 1/10
100/100 [==============================] - 12s 106ms/step - loss: 0.6981 - accuracy: 0.5265 - val_loss: 0.6923 - val_accuracy: 0.4980
Epoch 2/10
100/100 [==============================] - 12s 123ms/step - loss: 0.6924 - accuracy: 0.5365 - val_loss: 0.6842 - val_accuracy: 0.5560
Epoch 3/10
100/100 [==============================] - 12s 124ms/step - loss: 0.6750 - accuracy: 0.5745 - val_loss: 0.6660 - val_accuracy: 0.5880
Epoch 4/10
100/100 [==============================] - 10s 102ms/step - loss: 0.6712 - accuracy: 0.5745 - val_loss: 0.6882 - val_accuracy: 0.5630
Epoch 5/10
100/100 [==============================] - 12s 123ms/step - loss: 0.6634 - accuracy: 0.5900 - val_loss: 0.6632 - val_accuracy: 0.5840
Epoch 6/10
100/100 [==============================] - 12s 123ms/step - loss: 0.6462 - accuracy: 0.6225 - val_loss: 0.6440 - val_accuracy: 0.6260
Epoch 7/10
100/100 [==============================] - 10s 102ms/step - loss: 0.6491 - accuracy: 0.6080 - val_loss: 0.6492 - val_ac

Question 3

What is the median of training accuracy for this model?

Question 4

What is the standard deviation of training loss for this model?

In [53]:
history.history.keys()

dict_keys(['loss', 'accuracy', 'val_loss', 'val_accuracy'])

In [54]:
np.median(history.history['accuracy'])

0.5989999771118164

In [55]:
np.std(history.history['loss'])

0.02347979842346073

In [56]:
train_gen = ImageDataGenerator(rescale=1./255,
                               rotation_range=40,
                                width_shift_range=0.2,
                                height_shift_range=0.2,
                                shear_range=0.2,
                                zoom_range=0.2,
                                horizontal_flip=True,
                                fill_mode='nearest')

train_ds = train_gen.flow_from_directory(
                       '/content/train_data',
                        target_size=(150, 150),
                        batch_size=20
)

Found 20000 images belonging to 2 classes.


In [57]:
history = model.fit(
    train_ds,
    steps_per_epoch=100,
    epochs=10,
    validation_data=val_ds,
    validation_steps=50
)

Epoch 1/10
100/100 [==============================] - 21s 204ms/step - loss: 0.7043 - accuracy: 0.5000 - val_loss: 0.6625 - val_accuracy: 0.6640
Epoch 2/10
100/100 [==============================] - 20s 201ms/step - loss: 0.6950 - accuracy: 0.5000 - val_loss: 0.6727 - val_accuracy: 0.6560
Epoch 3/10
100/100 [==============================] - 22s 222ms/step - loss: 0.6940 - accuracy: 0.5000 - val_loss: 0.6787 - val_accuracy: 0.6150
Epoch 4/10
100/100 [==============================] - 20s 201ms/step - loss: 0.6936 - accuracy: 0.5000 - val_loss: 0.6820 - val_accuracy: 0.6170
Epoch 5/10
100/100 [==============================] - 20s 205ms/step - loss: 0.6934 - accuracy: 0.5000 - val_loss: 0.6858 - val_accuracy: 0.6070
Epoch 6/10
100/100 [==============================] - 22s 224ms/step - loss: 0.6933 - accuracy: 0.5000 - val_loss: 0.6859 - val_accuracy: 0.6090
Epoch 7/10
100/100 [==============================] - 20s 203ms/step - loss: 0.6933 - accuracy: 0.5000 - val_loss: 0.6862 - val_ac

Question 5

Let's train our model for 10 more epochs using the same code as previously. Make sure you don't re-create the model - we want to continue training the model we already started training.

What is the mean of validation loss for the model trained with augmentations?



Question 6


What's the average of validation accuracy for the last 5 epochs (from 6 to 10) for the model trained with augmentations?

In [58]:
np.mean(history.history['val_loss'])

0.6814465641975402

In [63]:
np.mean(history.history['val_accuracy'][5:])

0.6152000069618225